library/java/org/chromium/net/impl/CronetBidirectionalStream.java

```diff
@@ -233,259 +174,237 @@ public void run() {
   }
 
+  /**
+   * Saves the buffer intended to receive the data from the next read.
+   */
+  void recordReadBuffer(ByteBuffer buffer) {
+    mLatestBufferRead = buffer;
+    mLatestBufferReadInitialPosition = buffer.position();
+    mLatestBufferReadInitialLimit = buffer.limit();
+  }
+
   @Override
   public void write(ByteBuffer buffer, boolean endOfStream) {
-    synchronized (mNativeStreamLock) {
-      Preconditions.checkDirect(buffer);
-      if (!buffer.hasRemaining() && !endOfStream) {
-        throw new IllegalArgumentException("Empty buffer before end of stream.");
-      }
-      if (mEndOfStreamWritten) {
-        throw new IllegalArgumentException("Write after writing end of stream.");
-      }
-      if (isDoneLocked()) {
-        return;
-      }
-      mPendingData.add(buffer);
-      if (endOfStream) {
-        mEndOfStreamWritten = true;
-      }
+    Preconditions.checkDirect(buffer);
+    if (!buffer.hasRemaining() && !endOfStream) {
+      throw new IllegalArgumentException("Empty buffer before end of stream.");
+    }
+    if (mState.nextAction(endOfStream ? Event.USER_LAST_WRITE : Event.USER_WRITE) ==
+        NextAction.WRITE) {
+      mPendingData.add(new WriteBuffer(buffer, endOfStream));
     }
   }
 
   @Override
   public void flush() {
-    synchronized (mNativeStreamLock) {
-      if (isDoneLocked() ||
-          (mWriteState != State.WAITING_FOR_FLUSH && mWriteState != State.WRITING)) {
-        return;
-      }
-      if (mPendingData.isEmpty() && mFlushData.isEmpty()) {
-        // If there is no pending write when flush() is called, see if
-        // request headers need to be flushed.
-        if (!mRequestHeadersSent) {
-          mRequestHeadersSent = true;
-          CronetBidirectionalStreamJni.get().sendRequestHeaders(mNativeStream,
-                                                                CronetBidirectionalStream.this);
-          if (!doesMethodAllowWriteData(mInitialMethod)) {
-            mWriteState = State.WRITING_DONE;
-          }
-        }
-        return;
+    if (mUserflushConcurrentInvocationCount.getAndIncrement() > 0) {
+      // Another Thread is already copying pending buffers - can't be done concurrently.
+      // However, the thread which started with a zero count will loop until this count goes back
+      // to zero. For all intent and purposes, this has a similar outcome as using synchronized {}
+      return;
+    }
+    do {
+      WriteBuffer pendingBuffer;
+      while ((pendingBuffer = mPendingData.poll()) != null) {
+        mFlushData.add(pendingBuffer);
       }
-
-      assert !mPendingData.isEmpty() || !mFlushData.isEmpty();
-
-      // Move buffers from mPendingData to the flushing queue.
-      if (!mPendingData.isEmpty()) {
-        mFlushData.addAll(mPendingData);
-        mPendingData.clear();
+      if (mState.nextAction(Event.USER_FLUSH) == NextAction.FLUSH_HEADERS) {
+        mStream.sendHeaders(mEnvoyRequestHeaders, /* endStream= */ mReadOnly);
       }
```

Comment:
This will always be executed at most once, right?

If so, can you move this out of the loop?

Reply:
Yes - this statement can only be true no more than once. Changed - thanks!